File: python/day01.py
from aoc import *
from collections import Counter


def diff(a, b):
    return abs(a - b)

def part_1(columns):
    return sum(map(diff, *columns))

def part_2(l, r):
    freqs = Counter(r)
    return sum(n * freqs[n] for n in l)

def solve(filename):
    columns = mapl(sorted, transpose(parse_lines(read_input(filename), ints)))
    return part_1(columns), part_2(*columns)


solve(1)

File: elixir/day01.exs
defmodule Day01 do
  import Enum

  defp sort_cols(lines) do
    lines
    |> zip
    |> map(&sort(Tuple.to_list(&1)))
  end

  defp part_1(cols) do
    cols
    |> zip
    |> map(fn {x, y} -> abs(x - y) end)
    |> sum
  end

  defp part_2([left, right]) do
    freqs = frequencies(right)

    left
    |> map(fn x -> x * Map.get(freqs, x, 0) end)
    |> sum
  end

  def solve(filename) do
    import AOC

    filename
    |> read_input()
    |> parse_lines(:ints)
    |> sort_cols
    |> juxt([&part_1/1, &part_2/1])
  end
end

Day01.solve(1)
